A16-tcp-user-timeout.md

```diff
@@ -0,0 +1,55 @@
+
+At the time this proposal was written, gRPC Core faced an issue where data could be stuck for a very long time at the gRPC TCP layer. For cases where the network goes down, and the socket cannot accept more data for writing, an RPC could be blocked for a very long time. Even if the deadline expired on such an RPC, gRPC Core was incapable of shutting down the RPC, because the TCP layer would still have references to user data. Setting TCP_USER_TIMEOUT to a reasonable value would put a bound on the maximum time the TCP layer would wait for TCP packets to be acknowledged, and close the connection if the packets are not acknowledged in time. This would end up releasing references to user data at the gRPC TCP layer, thereby allowing the RPC to fail in a timely manner.
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+For all Linux platforms running Linux 2.6.37 and later, gRPC will set the socket option TCP_USER_TIMEOUT for TCP sockets to a default value of 20 seconds. (The system defaults might be as high as 20 minutes.)
+
+gRPC will also provide a configuration option for both channels and servers to set the TCP_USER_TIMEOUT value.
+
+## Rationale
+
+This proposal can help mitigate two separately seen issues, but with the same root cause that was described above -
+* [#15889](https://github.com/grpc/grpc/issues/15889) - gRPC would potentially block indefinitely and not respect RPC deadlines.
+* [#15871](https://github.com/grpc/grpc/issues/15871) - Heavy loads prevent keepalives from destroying the connection.
+Note that both issues occurred when a network was lost. The first issue mentioned in #15889 could also be solved by fixing the reference issue. This could be achieved by -
+* making a copy of the user data - This would result in a significant user cost
+* adding API to gRPC Core's TCP layer to be able to cancel writes for certain streams - This would require gRPC's TCP layer to understand streams and this breaks layers, not to mention the additional bookkeeping and overhead in doing so.
+
+Even by doing one of the two currently known solutions, the issue mentioned in #15871 would still remain, since we require sending the HTTP2 GOAWAY frame before closing the connection in certain scenarios.
+
+Other than solving the two described issues, TCP_USER_TIMEOUT would also provide a fairly reliable way to detect TCP connection breakage without any additional cost.
+
+## Implementation
+
+C-Core - [#16419](https://github.com/grpc/grpc/issues/16419) sets the TCP_USER_TIMEOUT socket option.
+A new channel argument, GRPC_ARG_TCP_USER_TIMEOUT_MS will be introduced to configure the value. The channel argument value will be passed as the socket option value. The minimum value, 0, would result in setting TCP_USER_TIMEOUT to the system default. The maximum value will be capped at 20 minutes (system default).
+
+JAVA - For JAVA, the support would be currently restricted to netty. The OkHttp library we use does not support setting such an option.
```

Comment:
Netty with NIO (which is our default) also doesn't support this. Only Netty with epoll/kqueue supports this, and when the user is configuring gRPC to use epoll/kqueue they can also configure TCP_USER_TIMEOUT via pass-through configuration in our API.

So as this proposal stands now, there no benefit to adding anything in Java.

Reply:
Thanks for checking! I'll update the doc to reflect this.